source/common/router/config_utility.cc

```diff
@@ -98,6 +99,34 @@ Optional<Http::Code> ConfigUtility::parseDirectResponseCode(const envoy::api::v2
   return Optional<Http::Code>();
 }
 
+std::string ConfigUtility::parseDirectResponseBody(const envoy::api::v2::Route& route) {
+  if (!route.has_direct_response() || !route.direct_response().has_body()) {
+    return EMPTY_STRING;
+  }
+  const auto& body = route.direct_response().body();
+  const std::string filename = body.filename();
+  if (!filename.empty()) {
+    static const ssize_t MaxFileSize = 4096;
+    if (!Filesystem::fileExists(filename)) {
+      throw EnvoyException(fmt::format("response body file {} does not exist", filename));
+    }
+    ssize_t size = Filesystem::fileSize(filename);
+    if (size < 0) {
+      throw EnvoyException(fmt::format("cannot determine size of response body file {}", filename));
+    }
+    if (size > MaxFileSize) {
+      throw EnvoyException(fmt::format("response body file {} size is {} bytes; maximum is {}",
+                                       filename, MaxFileSize));
+    }
+    return Filesystem::fileReadToEnd(filename);
+  }
+  const std::string inline_bytes = body.inline_bytes();
+  if (!inline_bytes.empty()) {
+    return inline_bytes;
+  }
+  return body.inline_string();
+}
```

Comment:
Sorry for being late to the party, but:
1. Why did everybody think that 4KB limit is a good idea?
2. Iff it's a good idea, then why does it only apply to response body from file, but not to response body inlined in the config?

cc @mattklein123 @alyssawilk

Reply:
The exact number was chosen by @brian-pane. He can comment. It's probably arbitrary. The main idea here is to provide some sanity check to avoid people from streaming huge files which won't reliably work without a lot of effort around buffering, flow control, etc. We should probably apply the same limit whether inline or by file I agree. @brian-pane can you do a follow up?

Reply:
@PiotrSikora the idea of using a fixed limit arose from a Slack discussion in #envoy-dev on Jan 5th. My original plan was to use sendfile, but Matt had (quite reasonable) concerns about how much complexity that would add to the proxy core, so we settled on reading the file into memory but limiting its size.

Reply:
BTW, PR [2458](https://github.com/envoyproxy/envoy/pull/2458) now contains a fix to apply the same 4KB limit to both inline and file-sourced response bodies.